Userland/Demos/CatDog/CatDog.h

```diff
@@ -25,112 +25,39 @@ class CatDog final : public GUI::Widget
 public:
     static ErrorOr<NonnullRefPtr<CatDog>> create();
 
-    // The general state, does not contain movement direction or whether CatDog is roaming.
-    enum class MainState {
-        Idle,      // default state
-        Alerted,   // woken by mouse cursor or speaking after being idle
-        Sleeping,  // mouse hasn't moved in some time
-        Artist,    // PixelPaint or FontEditor are open
-        Inspector, // SystemServer, Profiler or Inspector are open
-    };
-    static bool is_non_application_state(MainState state)
-    {
-        return state == MainState::Idle || state == MainState::Alerted || state == MainState::Sleeping;
-    }
-
-    virtual void timer_event(Core::TimerEvent&) override;
-    virtual void paint_event(GUI::PaintEvent& event) override;
-    virtual void track_mouse_move(Gfx::IntPoint point) override;
-    virtual void mousedown_event(GUI::MouseEvent& event) override;
-    virtual void context_menu_event(GUI::ContextMenuEvent& event) override;
-
-    void start_the_timer() { m_timer.start(); }
+    void timer_event(Core::TimerEvent&) final;
+    void paint_event(GUI::PaintEvent& event) final;
+    void track_mouse_move(Gfx::IntPoint point) final;
+    void mousedown_event(GUI::MouseEvent& event) final;
+    void context_menu_event(GUI::ContextMenuEvent& event) final;
 
     Function<void()> on_click;
     Function<void(GUI::ContextMenuEvent&)> on_context_menu_request;
 
-    bool roaming() const { return m_roaming; }
-    void set_roaming(bool roaming)
-    {
-        m_roaming = roaming;
-        if (!roaming) {
-            // If we stop CatDog while it's in a program-specific state, we don't want it to be alerted.
-            if (m_main_state == MainState::Idle || m_main_state == MainState::Sleeping)
-                m_main_state = MainState::Alerted;
-            m_curr_frame = 0;
-            set_image_by_main_state();
-            update();
-        }
-    }
+    void set_roaming(bool roaming);
 
-    MainState main_state() const { return m_main_state; }
+    [[nodiscard]] bool is_artist() const;
+    [[nodiscard]] bool is_inspector() const;
 
 private:
-    Gfx::IntPoint m_temp_pos;
-    Core::ElapsedTimer m_timer;
+    struct ImageForState {
+        uint16_t state;
+        RefPtr<Gfx::Bitmap> bitmap;
```

Comment:
Why store `RefPtr` here? And not `NonnullPtrRefPtr` since I would assume any image we add will be non null

Reply:
Used RefPtr because that's what was there.
But since this comes after the TRY() step now, I agree and changed it to Nonnull for the m_images vector. Done.